[quant][pt2e] Enable constant folding for quantize ops

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110125
* #110097
* __->__ #109343

Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

jerryzh168 committed:
[quant][pt2e] Enable constant folding for quantize ops

Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -78,6 +79,7 @@ def convert_pt2e(
 
     pm = PassManager([PortNodeMetaForQDQ()])
     model = pm(model).graph_module
+    constant_fold(model)
```
why are we constant folding everything? And why use util from inductor?

It seems pretty weird that this change is inside inductor. we should decouple from that or refactor inductor stuff into utils that can be used outside.

Plus, blanket const folding everything, including stuff not related to quantization that maybe there, does not seem right.

jerryzh168 replied:
yeah this is a bit tricky I feel. e.g. we could have weight -> transpose -> q -> dq as well, do you fold "transpose -> q" or not?

it is weird that we just want to constant fold quantize op and not others, I guess it seems clean to me to do this later in the step as well (in lowering) if it can not be done cleanly in the quantization step

jgong5 replied:
May I know the motivation of moving the constant folding to the quantization phase? Do we have a use case that one would save a quantized model and load it later for deployment? Then, the folding of weights to low-precision form can save the model size?

BTW, inductor's constant folding assumes that model params are marked as "freezing" with "get_attr" calls in the graph. Do we have those taken care of here too?

jerryzh168 replied:
yeah people are checking the size of the model after quantization + lowering to executorch and found that the size does not change right now

> inductor's constant folding assumes that model params are marked as "freezing" with "get_attr" calls in the graph. Do we have those taken care of here too?

not following this, could you clarify a bit?

jgong5 replied:
See here: https://github.com/pytorch/pytorch/blob/b3272b2c0081485140e0762023617cfa64a39c2b/torch/_inductor/freezing.py#L81

We need to know what inputs are model parameters according to the tracing context and then replace them with "get_attr" calls. Later on, the constant folding logic will work on those "get_attr" calls. Not sure what happens with the quantization path.

jerryzh168 replied:
here we just constant propagate `get_attr` nodes, it's not related to freezing I think

kimishpatel replied:
also what happens to metadata on the new get_attr nodes?

kimishpatel replied:
> yeah this is a bit tricky I feel. e.g. we could have weight -> transpose -> q -> dq as well, do you fold "transpose -> q" or not?

Jerry, yes you const fold through this. My point is handling only things related to quantization and not other things. If there were other transforms that user introduces before coming to quantization, making assumption on const folding those is not right.

jerryzh168 replied:
OK, sounds good, so the criteria is fold any constant that has a quantize op in the path then? I'd like to reuse inductor constant folding because it's the same thing and I don't want to reimplement it. I can see how I can refactor that to fit this requirement

jerryzh168 replied:
> also what happens to metadata on the new get_attr nodes?

good question, I'll check this

kimishpatel reviewed (changes requested):
I think we should const fold only things related to quant workflow. Plus inductor specific utils does not seem the right thing to use. We should probably factor those out if we want to use them here

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jerryzh168 commented:
@kimishpatel I saw metadata for the fused get_attr node is gone right now, but what should it be? e.g. in cases like (get_attr(weight) -> transpose -> quantize)

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

kimishpatel commented:
> @kimishpatel I saw metadata for the fused get_attr node is gone right now, but what should it be? e.g. in cases like (get_attr(weight) -> transpose -> quantize)

For the most part i would assume that it is same getween get_attr and tranpose. But I imagine porting metadata from tranpose node should be ok for now. Just add checks to ensure that all nodes between get_attr and quantize have the same source_fn and nn_module_stack, just so that we can catch such cases, when they arise, and think of the fix we need.

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -66,6 +69,22 @@ def prepare_qat_pt2e(
 
+_QUANT_OPS = [
+    torch.ops.quantized_decomposed.quantize_per_tensor.default,
+    torch.ops.quantized_decomposed.quantize_per_tensor.tensor,
```
this should include quantize_per_channel as well

jerryzh168 replied:
oh yeah I updated this locally let me update the PR

kimishpatel commented on torch/_inductor/constant_folding.py:
```diff
@@ -163,11 +167,13 @@ def run(self):
 
 @torch.utils._python_dispatch._disable_current_modes()
-def constant_fold(gm):
+def constant_fold(gm, filter_fn: Optional[Callable[[torch.fx.Node], bool]] = None):
```
Please have someone from inductor, who touched this file before, also review this part of the code.

kimishpatel commented on torch/_inductor/constant_folding.py:
```diff
@@ -51,7 +51,11 @@ def __init__(
+            torch.ops.quantized_decomposed.dequantize_per_channel.default,
+            torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+            torch.ops.quantized_decomposed.dequantize_per_tensor.tensor,
+        ]:
```
These changes look good!

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -66,6 +69,23 @@ def prepare_qat_pt2e(
+
+    Note: this filter does not handle the case when there are any additional ops after quantize:
+    e.g. get_attr(weight) -> transpose1 -> quantize -> transpose2 -> dequantize*
+    this filter will filter out the above example since the last op in the sequence and we'll not
```
This is because the const prop will be associated with "tranpose" node, for which this function will return false? Makes sense, but just trying to understand.

jerryzh168 replied:
yes, in this case when we replace node with constant Tensors, we'll be replacing transpose2 node with a constant Tensor and the check will fail

jgong5 replied:
May I know the motivation of this filtering? What's the problem if we fold the weights till `transpose2`?

jerryzh168 replied:
this is not a good example I think, the point is that we only fold things related to q/dq, not folding `transpose2` is an artifact of the current implementation details, it's not that important.

kimishpatel commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -78,6 +98,8 @@ def convert_pt2e(
     pm = PassManager([PortNodeMetaForQDQ()])
     model = pm(model).graph_module
+
+    constant_fold(model, _quant_node_filter)
```
Do we not want to hide this behind feature flag similar to `user_reference_representation`, because it may break some existing jobs. Or fix those perhaps?

jerryzh168 replied:
will fix the tests

kimishpatel replied:
this is not just the tests btw. There is xnnpack partitioner and other stuff elsewhere.

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +915,101 @@ def forward(self, x):
+        """Make sure the constant propagation does not apply to things unrelated to
+        quantization
+        """
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.linear = torch.nn.Linear(2, 2)
+                self.dont_fold_me = torch.randn(2, 2)
+
+            def forward(self, x):
+                t = self.dont_fold_me.t()
+                return self.linear(x) + t
```
You can just refactor this to take is non_const_fold op as arg, so that you can combine this with _get_pt2e_quantized_linear

jerryzh168 replied:
feels like this might make the original helper hard to read, so I'll keep this for now, maybe need to refactor the prepare -> calibrate -> convert code to a single call to deduplicate code

kimishpatel replied:
I wouldnt imagine it will be too hard to read. I just want to reduce boilerplate stuff. But not too opinionated here.

jerryzh168 replied:
maybe I can factor out the quantization part of code

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +915,101 @@ def forward(self, x):
+        m = self._get_pt2e_quantized_linear()
+        node_occurrence = {
+            # quantize op for weight node is folded
+            ns.call_function(torch.ops.quantized_decomposed.quantize_per_tensor.default): 2,
```
Please add per channel test as well

kimishpatel reviewed (changes requested):
Left some comments

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +917,113 @@ def forward(self, x):
+        m = self._get_pt2e_quantized_linear()
+        for n in m.graph.nodes:
+            if n.op == "get_attr" and "frozen_param" in n.target:
+                self.assertIn("stack_trace", n.meta)
```
we should record and compare against the original metadata.

kimishpatel commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -98,6 +98,19 @@ def _port_metadata_for_input_quant_nodes(
+            q_node_input = q_node_input.args[0]
+        if isinstance(q_node_input, torch.fx.Node) and q_node_input.op == "get_attr":
+            for n in q_to_get_attr_nodes:
+                _add_metadata(n, node)
```
I had a comment here which I dont see now, but you dont want to add metadata from the 'node'. one you need to port metadata from is get_attr. So for a pattern like `get_attr -> q -> dq -> linear` the q node should get metadata from get_attr and not linear node. This is also why in the test we should test for having ported the right metadata, not just existence of metadata

jerryzh168 replied:
oh I see, OK will fix

kimishpatel reviewed (changes requested):
some more comments to be addressed. also want to make sure Digant is in the loop as this affects et delegation

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

kimishpatel reviewed (changes requested):
Only one major comment to address. rest looks good. lets make sure this doesnt break delegation as that may cause revert of this one.

jerryzh168 commented on test/quantization/pt2e/test_duplicate_dq.py:
```diff
@@ -106,19 +112,9 @@ def _test_duplicate_dq(
-                                f"No dq node found for {n}, even though {n} annotated for quantization."
-                            )
-                        self.assertEqual(len(dq_node.users.keys()), 1)
+                for arg in n.args:
+                    if arg.target in _DEQUANTIZE_OPS:
+                        self.assertEqual(len(arg.users.keys()), 1)
```
this is because the old code doesn't work anymore since the weight node is removed because of const prop

jerryzh168 commented on test/quantization/pt2e/test_metadata_porting.py:
```diff
@@ -57,6 +57,7 @@ def _tag_partitions(
 }
 
 
+# TODO: rename to TestPortMetadataPass to align with the util name?
```
oh what I mean is the util name is port_metadata_pass, so the test should be aligned with that

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +919,129 @@ def forward(self, x):
+
+            def forward(self, x):
+                t = self.weight.t()
+                return torch.nn.functional.linear(x, t)
```
this is testing that we'll fold all ops before quantize op

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +919,129 @@ def forward(self, x):
+
+        quantizer = XNNPACKQuantizer()
+        operator_config = get_symmetric_quantization_config(is_per_channel=False)
+        # only quantize linear, so add add is not quantized and the constant Tensor
```
oh this is not related, probably messed up when copying things

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +919,129 @@ def forward(self, x):
+        """Test to make sure the get_attr node for const propagated weight Tensor gets the correct
+        metadata (from original get_attr node from weight)
+        """
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.linear = torch.nn.Linear(2, 2)
+
+            def forward(self, x):
+                return self.linear(x)
+
+        quantizer = XNNPACKQuantizer()
+        operator_config = get_symmetric_quantization_config()
+        quantizer.set_global(operator_config)
+        example_inputs = (torch.randn(2, 2),)
+        m = M().eval()
+        m = capture_pre_autograd_graph(
+            m,
+            example_inputs,
+        )
+        weight_meta = None
+        for n in m.graph.nodes:
+            if n.op == "get_attr" and list(n.users)[0].target == torch.ops.aten.linear.default:
+                weight_meta = n.meta
+                break
+        assert weight_meta is not None, "Expect to find metadata for weight node"
+
+        m = prepare_pt2e(m, quantizer)
+        m(*example_inputs)
+        m = convert_pt2e(m)
```
we need to get weight_meta after capture so can't replace it I think

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -892,6 +919,129 @@ def forward(self, x):
+
+        for n in m.graph.nodes:
+            if n.op == "get_attr" and "frozen_param" in n.target:
+                self.assertIn("stack_trace", n.meta)
```
see next line?

digantdesai commented on torch/_inductor/constant_folding.py:
```diff
@@ -163,11 +167,13 @@ def run(self):
 
 @torch.utils._python_dispatch._disable_current_modes()
-def constant_fold(gm):
+def constant_fold(gm, filter_fn: Optional[Callable[[torch.fx.Node], bool]] = None):
```
nit `filter_fn` is somewhat ambiguous as in filtering which way, `constraint_fn` might be better i.e only fold if following `constraint` is satisfied. Not a string opinion. Ok with filter_fn too.

jerryzh168 replied:
oh OK

digantdesai commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -66,6 +69,23 @@ def prepare_qat_pt2e(
+    torch.ops.quantized_decomposed.quantize_per_channel.default,
+]
+def _quant_node_filter(n: Node) -> bool:
+    """If there is any pure ops between get_attr and quantize op they will be const propagated
```
This comment can be improved. It is somewhat confusing

jerryzh168 replied:
which part is confusing? maybe I don't need to mention the second paragraph here?

jerryzh168 commented on torch/ao/quantization/pt2e/port_metadata_pass.py:
```diff
@@ -98,6 +98,19 @@ def _port_metadata_for_input_quant_nodes(
+            q_node_input = q_node_input.args[0]
+        if isinstance(q_node_input, torch.fx.Node) and q_node_input.op == "get_attr":
+            for n in q_to_get_attr_nodes:
+                _add_metadata(n, q_node_input)
```
I think we probably don't want to do it for the same reason that we don't inherit this for q/dq nodes? i.e. if nn_module_stack points to nn.Linear, the nodes here might actually be corresponding to {some quantized Linear module} instead

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

leslie-fang-intel commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -78,6 +109,10 @@ def convert_pt2e(
     pm = PassManager([PortNodeMetaForQDQ()])
     model = pm(model).graph_module
+
+    if fold_quantized_weight:
```
Is it a typo `fold_quantized_weight`? Looks like the input parameter is `fold_quantize`.

jerryzh168 replied:
oh right, I'll update this soon

leslie-fang-intel commented on test/quantization/pt2e/test_x86inductor_quantizer.py:
```diff
@@ -284,7 +284,8 @@ def test_conv2d_with_quantizer_api(self):
                 torch.ops.quantized_decomposed.dequantize_per_tensor.default: 2,
-                torch.ops.quantized_decomposed.quantize_per_channel.default: 1,
+                # note: quantize op for weights are const propagated
+                torch.ops.quantized_decomposed.quantize_per_channel.default: 0,
```
Hi @jerryzh168, another question is looks like `fold_quantize` doesn't turn on by default, why will these number got changed?

jerryzh168 replied:
this is the long term plan, so I'm setting this flag to True in x86 tests as well, I'll update the PR soon.

jerryzh168 replied:
forgot to test before updating the current commit

leslie-fang-intel replied:
Got it. Thanks for the detail explanation.

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

kimishpatel commented on torch/ao/quantization/quantize_pt2e.py:
```diff
@@ -74,22 +74,31 @@ def prepare_qat_pt2e(
+    Args:
+    model: calibrated/trained model
+    use_reference_representation: boolean flag to indicate whether to produce referece representation or not
+    fold_quantize: boolean flag to indicate whether fold the quantize op or not
```
Put a note saying, this will eventually be removed, so users should prefer using fold_quantize=True for future compatibility. And a note saying, there will be a BC breaking change in future.

kimishpatel reviewed (approved):
Do make sure that no internal workflows are breaking. I do see you added a flag, but just to be on safer side.

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
> Do make sure that no internal workflows are breaking. I do see you added a flag, but just to be on safer side.

yeah seems fine

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 commented:
@pytorchbot merge

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 commented:
@pytorchbot merge

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 committed:
Update on "[quant][pt2e] Enable constant folding for quantize ops"


Summary:
This PR added constant folding for quantize ops so that instead of storing fp32 weight in the
quantized model, we'll get int8/int16 etc. weight

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_fold_quantize

also will verify in executorch later

Reviewers:

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D49399210](https://our.internmc.facebook.com/intern/diff/D49399210)

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49399210).

jerryzh168 commented:
@pytorchbot merge
